common/src/message.rs

```diff
@@ -0,0 +1,488 @@
+                    OpCode::Drop => Some(Message::DropMessage { handle: left }),
+                    _ => unreachable!(),
+                }
+            }
+        }
+    }
+
+    fn msg_size(op: &OpCode) -> usize {
+        match op {
+            OpCode::CreateBlob => 16,
+            OpCode::CreateTree => 16,
+            OpCode::CreateThunk => 9,
+            OpCode::RunThunk => 9,
+            OpCode::Apply => 18,
+            OpCode::Reply => 9,
+            OpCode::Drop => 9,
+        }
+    }
+}
+
+struct MessageParser<'a> {
+    pending_opcode: Option<OpCode>,
+    pending_opcode_buf: [u8; 1],
+    pending_payload: [u8; 18],
+    pending_payload_offset: usize,
+    expected_payload_size: usize,
+    rb: RingBufferReceiver<'a>,
+}
+
+impl<'a> MessageParser<'a> {
+    fn new(rb: RingBufferReceiver<'a>) -> MessageParser<'a> {
+        Self {
+            pending_opcode: None,
+            pending_opcode_buf: [0; 1],
+            pending_payload: [0; 18],
+            pending_payload_offset: 0,
+            expected_payload_size: 0,
+            rb,
+        }
+    }
+
+    // Read exactly once
+    fn read(&mut self) -> Result<Option<Message>, RingBufferError> {
+        match &self.pending_opcode {
+            None => match self.rb.read(&mut self.pending_opcode_buf) {
+                Ok(_) => {
+                    self.pending_opcode = Some(
+                        OpCode::try_from(self.pending_opcode_buf[0])
+                            .expect("Failed to parse OpCode"),
+                    );
+                    self.expected_payload_size =
+                        Message::msg_size(&self.pending_opcode.as_ref().unwrap());
+                    Ok(None)
+                }
+                Err(e) => Err(e),
+            },
+            Some(op) => {
+                match self.rb.read(
+                    &mut self.pending_payload
+                        [self.pending_payload_offset..self.expected_payload_size],
+                ) {
+                    Ok(n) => {
+                        self.pending_payload_offset += n;
+                        // Check whether end of message
+                        if self.pending_payload_offset == self.expected_payload_size {
+                            let result = Message::parse(op, &self.pending_payload);
+
+                            // Reset variables
+                            self.pending_payload_offset = 0;
+                            self.pending_opcode = None;
+
+                            match result {
+                                Some(msg) => Ok(Some(msg)),
+                                None => Err(RingBufferError::ParseError),
+                            }
+                        } else {
+                            Ok(None)
+                        }
+                    }
+                    Err(e) => Err(e),
+                }
+            }
+        }
+    }
+
+    fn read_exact_one(&mut self) -> Result<Message, RingBufferError> {
```

Comment:
Maybe just `read_one` since it's not taking a mutable slice like `read_exact` usually does.